src/iqm/iqm_client/authentication.py

```
@@ -0,0 +1,303 @@
+
+        Raises ClientAuthenticationError if getting the token fails.
+        """
+
+        if self._token_provider is None:
+            return None  # Authentication is not used
+
+        # Use the existing access token if it is still valid
+        if self._access_token and TokenManager.time_left_seconds(self._access_token) > REFRESH_MARGIN_SECONDS:
+            return f'Bearer {self._access_token}'
+
+        # Otherwise, get a new access token from token provider
+        try:
+            self._access_token = self._token_provider.get_token()
+            return f'Bearer {self._access_token}'
+        except ClientAuthenticationError:
+            if retries < 1:
+                raise
+
+        # Try again
+        return self.get_bearer_token(retries - 1)
+
+    def close(self) -> bool:
+        """Close the configured token provider.
+
+        Returns:
+            True if closing was successful
+
+        Raises:
+            ClientAuthenticationError: if closing failed
+        """
+        if self._token_provider is None:
+            return False
+
+        self._token_provider.close()
+        self._token_provider = None
+        return True
+
+
+class TokenProviderInterface(ABC):
+    """Interface to token provider"""
+
+    @abstractmethod
+    def get_token(self) -> str:
+        """
+        Returns a valid access token.
+
+        Raises ClientAuthenticationError if acquiriung the token fails.
+        """
+
+    @abstractmethod
+    def close(self) -> None:
+        """Closes authentication session. Returns True if closing was successful"""
+
+
+class ExternalToken(TokenProviderInterface):
+    """Holds an external token"""
+
+    def __init__(self, token: str):
+        self._token: Optional[str] = token
+
+    def get_token(self) -> str:
+        if self._token is None:
+            raise ClientAuthenticationError('No external token available')
+        return self._token
+
+    def close(self) -> None:
```

~Is it supposed to raise an exception **every** time it gets called? It feels a bit counter-intuitive...~

---

Calling close is needed only when an auth session needs to be closed (by logout). There should be no need to call it multiple times.

---

Yes, understood, but the logout will end up in an exception thrown every single time in case e.g. a token from a file was used. That I find a bit counter-intuitive as that exception canot be avoided, it is "baked" into the design :-)

---

Typically the user never needs to call 'close'. It is called automatically when client's '__del__' is called, i.e. when the client is garbage collected, and there any exceptions are silently ignored.